lib/lambda_ethereum_consensus/state_transition/operations.ex

```diff
@@ -835,41 +814,74 @@ defmodule LambdaEthereumConsensus.StateTransition.Operations do
     end
   end
 
-  defp invalid_target_epoch?(data, state) do
-    data.target.epoch < Accessors.get_previous_epoch(state) ||
-      data.target.epoch > Accessors.get_current_epoch(state)
+  defp check_valid_target_epoch(data, state) do
+    if data.target.epoch in [
+         Accessors.get_previous_epoch(state),
+         Accessors.get_current_epoch(state)
+       ] do
+      :ok
+    else
+      {:error, "Invalid target epoch"}
+    end
   end
 
-  defp epoch_mismatch?(data) do
-    data.target.epoch != Misc.compute_epoch_at_slot(data.slot)
+  defp check_epoch_matches(data) do
+    if data.target.epoch == Misc.compute_epoch_at_slot(data.slot) do
+      :ok
+    else
+      {:error, "Epoch mismatch"}
+    end
   end
 
-  defp invalid_slot_range?(data, state) do
-    state.slot < data.slot + ChainSpec.get("MIN_ATTESTATION_INCLUSION_DELAY") ||
-      state.slot > data.slot + ChainSpec.get("SLOTS_PER_EPOCH")
+  defp check_valid_slot_range(data, state) do
+    if data.slot + ChainSpec.get("MIN_ATTESTATION_INCLUSION_DELAY") <= state.slot and
+         state.slot <= data.slot + ChainSpec.get("SLOTS_PER_EPOCH") do
+      :ok
+    else
+      {:error, "Invalid slot range"}
+    end
   end
 
-  defp exceeds_committee_count?(data, state) do
-    data.index >= Accessors.get_committee_count_per_slot(state, data.target.epoch)
+  defp check_committee_count(data, state) do
+    if data.index >= Accessors.get_committee_count_per_slot(state, data.target.epoch) do
+      {:error, "Index exceeds committee count"}
+    else
+      :ok
+    end
   end
 
-  defp mismatched_aggregation_bits_length?(attestation, beacon_committee) do
-    length_of_bitstring(attestation.aggregation_bits) - 1 != length(beacon_committee)
+  defp check_matching_aggregation_bits_length(attestation, beacon_committee) do
+    if length_of_bitlist(attestation.aggregation_bits) == length(beacon_committee) do
+      :ok
+    else
+      {:error, "Mismatched aggregation bits length"}
+    end
   end
 
-  defp valid_signature?(state, indexed_attestation) do
-    Predicates.is_valid_indexed_attestation(state, indexed_attestation)
+  defp check_valid_signature(state, indexed_attestation) do
+    if Predicates.is_valid_indexed_attestation(state, indexed_attestation) do
+      :ok
+    else
+      {:error, "Invalid signature"}
+    end
   end
 
-  defp length_of_bitstring(binary) when is_binary(binary) do
-    binary
-    |> :binary.bin_to_list()
-    |> Enum.reduce("", fn byte, acc ->
-      acc <> Integer.to_string(byte, 2)
-    end)
-    |> String.length()
+  defp length_of_bitlist(bitlist) when is_binary(bitlist) do
+    bit_size = bit_size(bitlist)
+    <<_::size(bit_size - 8), last_byte>> = bitlist
+    bit_size - leading_zeros(<<last_byte>>) - 1
   end
 
+  defp leading_zeros(<<1::1, _::7>>), do: 0
```

Comment:
maybe move to Utils.BitVector?

Reply:
It's kind of unrelated. I was gonna move the whole "compute bitlist's length" function to a new `Utils.BitList` module, but I think it's better to tackle that in a future refactor.